ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs

```diff
@@ -0,0 +1,482 @@
+defaultGenTestConfig adhocTestConfig numCoreNodes numSlots = do
+    nodeJoinPlan <- genNodeJoinPlan numCoreNodes numSlots
+    nodeTopology <- genNodeTopology numCoreNodes
+    pure TestConfig
+      { adhocTestConfig
+      , nodeJoinPlan
+      , nodeTopology
+      , numCoreNodes
+      , numSlots
+      }
+
+-- | Shrink, including the number of nodes and slots
+--
+shrinkTestConfigFreely :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigFreely TestConfig
+  { adhocTestConfig
+  , numCoreNodes
+  , numSlots
+  , nodeJoinPlan
+  , nodeTopology
+  } =
+    dropId $
+    [ TestConfig
+        { adhocTestConfig
+        , nodeJoinPlan    = p'
+        , nodeTopology    = top'
+        , numCoreNodes    = n'
+        , numSlots        = t'
+        }
+    | n'             <- andId shrink numCoreNodes
+    , t'             <- andId shrink numSlots
+    , let truncateP   = truncateNodeJoinPlan nodeJoinPlan n' (numSlots, t')
+    , let truncateTop = truncateNodeTopology nodeTopology n'
+    , p'             <- andId shrinkNodeJoinPlan truncateP
+    , top'           <- andId shrinkNodeTopology truncateTop
+    ]
+
+-- | Shrink, including the number of slots but not number of nodes
+--
+shrinkTestConfigSlotsOnly :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigSlotsOnly TestConfig
+  { adhocTestConfig
+  , numCoreNodes
+  , numSlots
+  , nodeJoinPlan
+  , nodeTopology
+  } =
+    dropId $
+    [ TestConfig
+        { adhocTestConfig
+        , nodeJoinPlan    = p'
+        , nodeTopology    = top'
+        , numCoreNodes
+        , numSlots        = t'
+        }
+    | t'             <- andId shrink numSlots
+    , let truncateP   = truncateNodeJoinPlan nodeJoinPlan numCoreNodes (numSlots, t')
+    , let truncateTop = truncateNodeTopology nodeTopology numCoreNodes
+    , p'             <- andId shrinkNodeJoinPlan truncateP
+    , top'           <- andId shrinkNodeTopology truncateTop
+    ]
+
+truncateNodeJoinPlan ::
+    NodeJoinPlan -> NumCoreNodes -> (NumSlots, NumSlots) -> NodeJoinPlan
+truncateNodeJoinPlan
+  (NodeJoinPlan m) (NumCoreNodes n') (NumSlots t, NumSlots t') =
+    NodeJoinPlan $
+    -- scale by t' / t
+    Map.map (\(SlotNo i) -> SlotNo $ (i * toEnum t') `div` toEnum t) $
+    -- discard discarded nodes
+    Map.filterWithKey (\(CoreNodeId nid) _ -> nid < n') $
+    m
+
+truncateNodeTopology :: NodeTopology -> NumCoreNodes -> NodeTopology
+truncateNodeTopology (NodeTopology m) (NumCoreNodes n') =
+    NodeTopology $ Map.filterWithKey (\(CoreNodeId i) _ -> i < n') m
+
+{-------------------------------------------------------------------------------
+  MockBFT test
+-------------------------------------------------------------------------------}
+
+type BlockMockBFT = SimpleBftBlock SimpleMockCrypto BftMockCrypto
+
+instance TestableBlock BlockMockBFT where
+    type AdhocTestConfig BlockMockBFT = SecurityParam
+
+    genTestConfig = join $
+      defaultGenTestConfig (SecurityParam 5)
```

Comment:
It's tough to say. I would guess "Yes, but it might just be an uninteresting hiccup".

I know that protocols with anachrony limits (just `RealPBFT` right now) may fail in various ways depending on the `k/N/t` ratio (where `t` is the PBFT signature threshold). But in production we always have `k >> N`, right?

We also use a relatively small number of slots `S` in each test run (`S <= 100`, right now); maybe large `S/k` ratios might challenge some assumptions.

Reply:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.